src/DDMinimizer.cpp

```diff
@@ -0,0 +1,364 @@
+#include "DDMinimizer.hpp"
+
+#include "CircuitOptimizer.hpp"
+#include "CircuitSimulator.hpp"
+#include "QuantumComputation.hpp"
+#include "operations/OpType.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <numeric>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace qc;
+using namespace std;
+
+namespace qc {
+
+/***
+ * Public Methods
+ ***/
+void optimizeInputPermutation(qc::QuantumComputation qc) {
+  qc::Permutation perm = DDMinimizer::createGateBasedPermutation(qc);
+  qc.initialLayout = perm;
+  qc::CircuitOptimizer::elidePermutations(qc);
+}
+
+qc::Permutation
+DDMinimizer::createGateBasedPermutation(qc::QuantumComputation& qc) {
+  std::map<std::string, std::map<std::pair<Qubit, Qubit>, int>> maps =
+      DDMinimizer::makeDataStructure(qc).first;
+  std::map<std::string, std::vector<int>> indices =
+      DDMinimizer::makeDataStructure(qc).second;
+  std::size_t bits = qc.getNqubits();
+
+  // iterate over all the ops and mark the index of the found x-c pairs in the
+  // map.
+  int instruction_index = 0;
+  for (const auto& op : qc.ops) {
```

Comment:
```suggestion
  for (const auto& op : qc) {
```
Just works and you do not need access to protected members ;-)

Reply:
That's perfect, thank you so much!